include/ocpp/v201/ocpp_types.hpp

```diff
@@ -959,6 +969,86 @@ void from_json(const json& j, Firmware& k);
+
+    ocpp::DateTime start;
+    ocpp::DateTime end;
+    float limit;
+    std::optional<std::int32_t> number_phases;
+    std::optional<std::int32_t> phase_to_use;
+    std::int32_t stack_level;
+    ChargingRateUnitEnum charging_rate_unit;
+    std::optional<float> min_charging_rate;
+
+    bool equals(const period_entry_t& other) const {
+        return (start == other.end) && (end == other.end) && (limit == other.limit) &&
+               (number_phases == other.number_phases) && (stack_level == other.stack_level) &&
+               (charging_rate_unit == other.charging_rate_unit) && (min_charging_rate == other.min_charging_rate);
+        // Add comparisons for other fields as necessary
+    }
+};
+
+bool operator==(const period_entry_t& a, const period_entry_t& b);
+bool operator!=(const period_entry_t& a, const period_entry_t& b);
+bool operator==(const std::vector<period_entry_t>& a, const std::vector<period_entry_t>& b);
+
+std::string to_string(const period_entry_t& entry);
+
+std::ostream& operator<<(std::ostream& os, const period_entry_t& entry);
+
+/// \brief calculate the start times for the profile
+/// \param now the current date and time
+/// \param end the end of the composite schedule
+/// \param session_start optional when the charging session started
+/// \param profile the charging profile
+/// \return a list of the start times of the profile
+std::vector<DateTime> calculate_start(const DateTime& now, const DateTime& end,
+                                      const std::optional<DateTime>& session_start, const ChargingProfile& profile);
+
+/// \brief calculate the start times for the schedule period
+/// \param now the current date and time
+/// \param end the end of the composite schedule
+/// \param session_start optional when the charging session started
+/// \param profile the charging profile
+/// \param period_index the schedule period index
+/// \return the list of start times
+std::vector<period_entry_t> calculate_profile_entry(const DateTime& now, const DateTime& end,
+                                                    const std::optional<DateTime>& session_start,
+                                                    const ChargingProfile& profile, std::uint8_t period_index);
+
+/// \brief generate an ordered list of valid schedule periods for the profile
+/// \param now the current date and time
+/// \param end ignore entries beyond this date and time (i.e. that start after end)
+/// \param session_start optional when the charging session started
+/// \param profile the charging profile
+/// \return a list of profile periods with calculated date & time start and end times
+/// \note it is valid for there to be gaps (for recurring profiles)
+std::vector<period_entry_t> calculate_profile(const DateTime& now, const DateTime& end,
+                                              const std::optional<DateTime>& session_start,
+                                              const ChargingProfile& profile);
```

Comment:
`calculate_profile()` is a bit of a vague name. It doesn't tell me that we're generating a list of valid schedule periods. Perhaps `calculate_valid_schedules_for_profile()`

Reply:
Understood. The idea is to match the method names from 1.6 as much as possible. Then, at a later date do a PR for both 1.6. and 2.0.1 to clean them up.